### PR TITLE
fix: reject not supported codecs

### LINKED ISSUE #2154: pinning service API needs to reject IPLD codecs we don’t support
- pinning service API needs to reject IPLD codecs we don’t support (for both web3.storage and NFT.Storage)
-  document that we support remote pinning the following codecs raw dag-json dag-cbor dag-pb in the root node and outside of that, we’ll reject (here https://web3.storage/docs/how-tos/pinning-services-api/ and https://nft.storage/docs/how-to/pinning-service/)
- also add to the API reference docs of each
- easiest way we can get this enforced in the API for now is to return non-200 responses to the pinning service API  (decode the cid in full, get its codec, compare, reject)

### PR DESCRIPTION
**Pull request by flea89**
Resolves #2154

### CODE REVIEW

File: packages/api/test/pin.spec.js [hunk 1]
```diff
@@ -562,6 +568,53 @@ describe('Pinning APIs endpoints', () => {
       assert.strictEqual(error.details, INVALID_CID)
     })
 
+    it('throws error if not supported codec', async () => {
+      const hash = await sha256.digest(Buffer.from(`${Math.random()}`))
```
Comment (francois-potato) [inline packages/api/test/pin.spec.js:572]: Nit: could use `toString()` to here too. (same line 599)

File: packages/api/src/utils/psa.js [hunk 2]
```diff
@@ -132,11 +149,25 @@ export function transformAndValidate (payload) {
   // Validate CID.
   if (cid) {
     try {
-      normalizedCid = normalizeCid(cid)
-      opts.cid = cid
+      let c
```
Comment (francois-potato) [inline packages/api/src/utils/psa.js:152]: Nit; I find the following code not easy to read. Not sure why the declaration of c is made here and not in the block below. Could we use only one try? Maybe define a type for the returned struct?

File: packages/api/test/pin.spec.js [hunk 3]
```diff
@@ -562,6 +568,53 @@ describe('Pinning APIs endpoints', () => {
...
+      const cid = CID.create(1, 72, hash).toString()
+
+      const res = await fetch(new URL('pins', endpoint).toString(), {
```
Comment (francois-potato) [inline packages/api/test/pin.spec.js:575]: Nit: fetch can use a URL object (no need to `toString`). (same line 603)

**Review by francois-potato — APPROVED** [review]
LGTM

File: packages/api/src/utils/psa.js [hunk 4]
```diff
@@ -132,11 +149,25 @@ export function transformAndValidate (payload) {
...
+        normalizedCid = c.toV1().toString()
+        opts.cid = cid
+      } catch (e) {
+        throw new PSAErrorInvalidData(INVALID_CID)
+      }
+
+      if (!SUPPORTED_CODECS.includes(c.code)) {
+        let message = PSACodecNotSupported.MESSAGE
+        if (c.code === 72) {
+          message = `${message} If you're trying to pin using an IPNS record that isn't supported yet. Please +1 [this githib issue](https://github.com/web3-storage/web3.storage/issues/2155) if you want it to be.`
```
Comment (blucas) [inline packages/api/src/utils/psa.js:164]:
```suggestion
          message = `${message} If you're trying to pin using an IPNS record that isn't supported yet. Please +1 [this github issue](https://github.com/web3-storage/web3.storage/issues/2155) if you want it to be.`
```
Does the client parsing this error message support markdown syntax for hyperlinks?

File: packages/api/src/utils/psa.js [hunk 5]
```diff
@@ -132,11 +149,25 @@ export function transformAndValidate (payload) {
...
+        c = CID.parse(cid)
+        normalizedCid = c.toV1().toString()
+        opts.cid = cid
+      } catch (e) {
+        throw new PSAErrorInvalidData(INVALID_CID)
+      }
+
+      if (!SUPPORTED_CODECS.includes(c.code)) {
+        let message = PSACodecNotSupported.MESSAGE
+        if (c.code === 72) {
```
Comment (blucas) [inline packages/api/src/utils/psa.js:163]: Should `72` be documented somewhere to identify its meaning?

Comment (codeflyer) [reply packages/api/src/utils/psa.js:163]: I'll add it to a constant, to improve the readability.

Comment (flea89) [reply packages/api/src/utils/psa.js]: Good point, saved it in a constant and added a bit of explanation.

File: packages/api/src/utils/psa.js (hunk 4 above)
Comment (flea89) [reply packages/api/src/utils/psa.js]: Good point, while obv this API can be used directly, I think it's safe to expect most of the requests would probably come from Kubo. And I don't know if it does support it.
I wonder if it's a real problem thogh. Even if it doesn't, Markdown syntax for links it's quite common anyway, and I expect users to be able to still understand and navigate to the issue.

What do you think?

**Review by codeflyer — APPROVED** [review]
LGTM

**Review by adamalton — APPROVED** [review]

**Review by alanshaw — APPROVED** [review]

**Pull request merged**